src/pip/_internal/resolution/resolvelib/candidates.py

```diff
@@ -68,13 +68,20 @@ def make_install_req_from_dist(dist, parent):
 
 
 class LinkCandidate(Candidate):
-    def __init__(self, link, parent, factory):
-        # type: (Link, InstallRequirement, Factory) -> None
+    def __init__(
+        self,
+        link,          # type: Link
+        parent,        # type: InstallRequirement
+        factory,       # type: Factory
+        name=None,     # type: Optional[str]
+        version=None,  # type: Optional[_BaseVersion]
+    ):
+        # type: (...) -> None
         self.link = link
         self._factory = factory
         self._ireq = make_install_req_from_link(link, parent)
-        self._name = None  # type: Optional[str]
-        self._version = None  # type: Optional[_BaseVersion]
+        self._name = name
```

Comment:
Just realised the name here might not be canonicalised properly. I’m including a path in my refactor PR for this.

Reply:
I was assuming that `InstallationCandidate` should have canonicalised. I didn't actually check, but I'm a little concerned that we're repeatedly re-normalising names, which may only be a small performance hit, but nevertheless it is a hit. And TBH it's probably easier to fix a bug where we don't canonicalise properly than it is to after the fact remove redundant calls to `canonicalize_name`.

I'm not going to object to you adding the call, but I do think it's a technical debt we should try to avoid. I'll raise an issue for it as a reminder.

Reply:
I went digging into the package finder, and your assumption is correct. I’ll drop the extra `canonicalize_name()` call for now, since it shouldn’t be a part of a refactoring PR anyway :p